### PR TITLE
Generalising class weights for all label ranks

### PR DESCRIPTION
**Pull request by j-bernardi**
Currently, model.fit(class_weight={0: 1.0, ...}) only works if y_data is of
rank 2, e.g. (batch_size, label_dim).

This change allows any label shape to have its outputs weighted,
e.g. image or temporal / sequence data. The relevant tests
are included.

Addresses issue #40457

### CODE REVIEW

**Comment by j-bernardi** [comment]
have done some local tests but waiting on a few more before requesting review

**Comment by j-bernardi** [comment]
I'd like to add to the tests. They should check that the data handler does actually produce logits that are adjusted by the class and sample weights as expected. Still in progress for this reason.

If there are additional suggestions / locations I should add unittests, please let me know. This is my first contribution so I'm unfamiliar with the test structure, despite some searching for the right places.

I think doing the above should give reasonable assurance that the feature is working, though, as the DataHandler directly serves the elsewhere-tested model.fit() function.

**Comment by j-bernardi** [comment]
ah-ha, I have discovered the rub with 3+ dimensional class weightings - there is no way to know whether the final dimension is one-hot encoded or a multidimensional output. I will spend a little time thinking but I think there would have to be some structural assertions that I'm not sure we're looking to make.

It could be inferred from number of classes in class_weights provided, but that would cause problems if the a flattened output happened to have the same number of dimensions as output classes

**Comment by j-bernardi** [comment]
I think this is ready and tested. I'll wait to see if unittests pass, then move to ready for review.

I'm getting a strange failure on the log caching locally, where
new_func_graph = "INFO:absl:tensorflow:Creating new FuncGraph for Python function" is not being found in the evaluate logs
which is fixed if I change to:
new_func_graph = "Level 1:tensorflow:Creating new FuncGraph for Python function"

I don't think I changed anything, so hopefully it just won't happen on this end.

**Comment by gbaned** [comment]
@j-bernardi  Can you please resolve conflicts? Thanks!

**Comment by ALIANS-PRODUCTIONS** [comment]
Hi Bernardi. I was also follow your thread here for the issue: Generalising class weights for all label ranks.
One of the main reasons why i wanted to use the latest tf-nightly builds.
Currently i cannot get class weights to work for multi label data (crash/error out for labels 3 or more)
Would be great to get your working implementation running.

**Comment by j-bernardi** [comment]
@gbaned  sure - just commenting to say it's not dead, but I haven't got around to it! 

Cheers

**Comment by j-bernardi** [comment]
Oops. Will look into it.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by j-bernardi** [comment]
Yes intending to work on this change still, apologies for the stall for anyone following.

**Comment by j-bernardi** [comment]
Converted to draft. Still intending to make the PR, but I'm struggling to get tests running on my new machine and not sure what led to the most recent failures.

**Comment by j-bernardi** [comment]
Managed to get the tests running locally again. Let's see if the checks turn anything up, then ready for re-review

File: tensorflow/python/keras/engine/data_adapter.py
Comment (fchollet) [inline tensorflow/python/keras/engine/data_adapter.py:1114]: I would recommend dropping this argument, and simply assuming that the class weights apply to the labels encoded along the last axis.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. I'll be taking over the review process. The feature looks like a very good idea. I don't think we need to gate it behind an argument.

File: tensorflow/python/keras/engine/data_adapter.py
Comment (j-bernardi) [reply tensorflow/python/keras/engine/data_adapter.py:1114]: Thanks for picking up this review - looking forward to working with you.

The reason I opted for the argument is that I believe making that assumption would break existing pipelines that are relying on the following current behaviour:
Currently, `fit` will interpret a final label dimension of size 1 or N_classes [if the dimension is <=2]. Your suggestion locks that down to requiring the final dim to be of size N_classes if class weights are being used, thus invalidating the former behaviour, right?

Assuming that the labels must be one-hot encoded (the N_classes-final-dim suggestion) does however resolve another problem, where it'd be ambiguous if final-dim size == N_classes, by coincidence. So it seems the sensible way to go to pick one.

This is all fine with me if it doesn't go against any design principles. I will make the change, just want to make the implications clear!

Comment (j-bernardi) [reply tensorflow/python/keras/engine/data_adapter.py:1114]: Another consideration: enforcing one-hot representation of labels means that people who depend on running class weights with memory-light, dense label representations will now have to use label tensors that are a factor 'n_classes' larger. Is that acceptable, and am I interpreting your change request correctly?

If we do not assume one-hot representation and do not ask for an argument flag, I think we are back to square 1 on why this feature was never implemented: it's ambiguous to coerce the label format, from the shape of the label alone. This would specifically fail in the case where the final dim size == n_classes, by coincidence. That'd be interpreted as 1-hot when the label was actually dense.

**Comment by gbaned** [comment]
@j-bernardi Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@j-bernardi Any update on this PR? Please. Thanks!

**Comment by j-bernardi** [comment]
@gbaned still on the todo list, if you're happy to leave it open for now. Thanks for continued interest in this change, to all involved.

I noticed the last response to reviewer requests went unreviewed from April 29th until the merge conflict on 25th June. I appreciate the reviewer likely had a lot on their plate(!), however I wanted to check I followed the correct process for requesting re-review, or if there's anything else I can do to help get this in? Thanks again!

**Comment by gbaned** [comment]
@j-bernardi Sorry for the delay. Once conflicts are resolved, we will try to review this PR as early as possible.  Thank you.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by j-bernardi** [comment]
I had intended to get back to this, but have not found the time. Converting to draft and I will re-request review once I get onto it, perhaps next month, if that's ok.

Cheers all

**Comment by fchollet** [comment]
@j-bernardi sorry that we weren't able to review this PR in a timely fashion. When you want to work on it again, please open the PR at [keras-team/keras](https://github.com/keras-team/keras) instead, which is our new development location. Over there we should be able to provide a timely review. Improve the contributor experience was the entire motivation for moving to a separate pure-Python repo.

**Pull request closed without merging**
